src/typechecker/typesystem.cpp

```diff
@@ -253,8 +254,8 @@ bool TypeSystemCore::ll_occurs(int v, int i){
 
 int TypeSystemCore::ll_find(int i) {
 	if (ll_node_header[i].tag == Tag::Term) return i;
-	if (ll_node_header[i].data_idx == i) return i;
-	return ll_node_header[i].data_idx = ll_find(ll_node_header[i].data_idx);
+	if (m_substitution[ll_node_header[i].data_idx] == -1) return i;
+	return ll_find(m_substitution[ll_node_header[i].data_idx]);
```

Comment:
This call to ll_find is unnecessary. Once a variable is substituted, it cannot change.